### ISSUE TITLE
Add background music

### ISSUE THREAD

**Issue by McPringle** [zdila/gpx-animator#35]
Add a music file (e.g. MP3 format) as a background music track for the animation. Fade out at the end of the animation.

**Comment by zdila** [comment]
I think this should be done in video editor software where you can interactively synchronize music to the video and adjust fade in/out times.

**Comment by zdila** [comment]
But this can make things complicated:

* it adds many more options to the app loosing focus to actial GPX animation
* different users can ask for different features - like showing subtitles at end of the viceo (color of the subtitles, animation styles like "StarWars", photo slideshow on the end or beginning of the video, special transitions between video parts etc)

I've created some videos using GPC animator without visibly loosing the quality. Please check https://www.youtube.com/watch?v=4ZM9lI1GppA.

Maybe what you actually want is to have GPX animator as a pugin for video editing woftware (like OpenShot, Kdenlive etc.).

**Comment by zdila** [comment]
Still adding just a music track could be acceptable. Even that you can add music to existing video later, using eg ffmpeg, without loosing anything from video quality (using copy codec).

**Comment by thomergil** [comment]
At the risk of being a repetitive nuisance, I think this is feature creep. I second @zdila  in that this should be done by whatever movie editing software that can take the output of GPX Animator as input. Alternatively, adding a soundtrack to video is also very easy to do this with a tool like `ffmpeg`.

**Comment by McPringle** [comment]
I often share animated routes that I created with GPX Animator. With these, I always have subtle music in the background. It annoys me to have to make only for this always another step over a video editing program with import of video and music, mix together and then again an export. I would like a file selection dialog in which you can select an MP3 file and a slider for the volume. This is then only two clicks for me in the future instead of having to do that extra in another program.

**Comment by zdila** [comment]
If it is easy to implement then I think it would be a nice feature despite of my previous comment :-). But then we may be tempted to solve volume fading in/out on start/end, ...

**Comment by thomergil** [comment]
I respectfully disagree with this argument. A lot of things are easy to implement, but that doesn't mean they make sense. (Especially if there are other tools that are better at performing the task.) It's probably also easy to implement a feature that rotates the entire video 180º. Yet we all agree that `ffmpeg` would be more suited for that. I feel the same argument holds for adding audio.

**Comment by zdila** [comment]
My reasons why it should be there:

- it is uncommon to have a video without a sound
- gpx animator, since it has GUI, is targeting also ordinary users. They know nothing about ffmpeg and probably doesn't have time and will to google and learn it (where they would need to use commandline). They want one tool to make their job (ie. make video from their GPX, naturally with audio)
- from UX perspective this will add only one field and one commandline option, thus not making UI any measurably more complex
- if rotatig a video of 180 degrees would be a very common task (more than 50% cases) then gpx animator could handle it too. But in my opinion it is barely more than 1% ;-).

**Comment by thomergil** [comment]
@zdila, I won't belabor the point, and I don't feel too strongly. I guess this comes down to a different opinion about what you and I refer to as "common" and "uncommon".

This feature comes with a few questions. You already pointed some out. Should music fade in? And what if the audio is longer than the video? Should it cut off? Should it fade out? How aggressively should it fade out? Should the length of the fade-out depend on the length of the full video? What if the video is only 10 seconds? Should it fade out for 1 second or 5 seconds? Should fade out be linear or non-linear? If the audio is shorter than the video? Can I add multiple songs to one video? How do I select them? Should all the audio be in the same format? How does one song blend into another? Using blending? Fading? Or just a hard stop/start? Should there be a pause between songs? How long? Is that configurable? What formats should it support? MP3? FLAC? Ogg Vorbis? WAV? At what audio quality? Should GPX Animator convert audio formats? Why not, if it's made for people that don't know about `ffmpeg`?

**Comment by zdila** [comment]
It should have some defaults. For advanced uses user would need to use ffmpeg (which would not be often).

- no fade in on beginning
- fade out - for example flashback duration. Or just 2 seconds. Linear. No rocket sciece ;-)
- single audio, non-loopable
- support mp3 or any format which will be supported by implementing library
- for all other question - just reasonable nonconfigurable default

**Comment by McPringle** [comment]
I have the same opinion as @zdila. I would really love this feature myself, as I explained above. With a very simple solution, 95% of the users of this feature would be satisfied.

- no fade in on the beginning -> every song has a beginning, so we don't need to fade something in
- if the music is longer than the video, linear fade out with a fixed length (can be made configurable later, if needed)
- one music file in MP3 format - if the user wants more, he can use other tools as ffmpeg or Windows Movie Maker...
- two configuration options: music file and volume

With the total time setting the length of the video can be aligned to the music, if necessary. Easy solution, easy to implement, won't "pollute" GPX Animator much, and is a nice to have feature.

**Comment by thomergil** [comment]
Sounds good. No pun intended.

**Comment by onixred** [comment]
add pull request https://github.com/gpx-animator/gpx-animator/pull/783
